src/objective_functions/kl_divergence.cpp

```diff
@@ -0,0 +1,133 @@
+
+void kl_divergence::setup(model& m) {
+  objective_function_term::setup(m);
+  //set up layers of interest
+  for(const auto& l : m.get_layers()) {
+    if(l->get_name() == m_z_mean_layer_name) m_z_mean_layer = l; 
+    if(l->get_name() == m_z_log_sigma_layer_name) m_z_log_sigma_layer = l;
+  }
+  
+  if (m_z_mean_layer == nullptr || m_z_log_sigma_layer == nullptr)
+  {
+    std::stringstream err;
+    err << __FILE__ << " " << __LINE__ << " :: "
+        << "Error: Null pointer(s) to layer(s)";
+    throw lbann_exception(err.str());
+  }
+}
+
+EvalType kl_divergence::evaluate() {
+
+  //Get matrices of input (latent space) layers
+  AbsDistMat& z_mean_acts = m_z_mean_layer->get_activations();
+  AbsDistMat& z_log_sigma_acts = m_z_log_sigma_layer->get_activations();
+  
+  AbsDistMat* z_mean = z_mean_acts.Construct(z_mean_acts.Grid(),
+                                             z_mean_acts.Root());
+  AbsDistMat* z_log1 = z_log_sigma_acts.Construct(z_log_sigma_acts.Grid(),
+                                                     z_log_sigma_acts.Root());
+
+  DataType mean_value = DataType(0); 
+  DataType std_value = DataType(0);
+  
+  auto sq = [&](const DataType& x) {
+    return (x*x);
+  };
+  auto expn  = [&](const DataType& x) {
+    return std::exp(x);
+  };
+  auto addone  = [&](const DataType& x) {
+    return (1+x);
+  };
+  
+  //Local copies
+  El::Copy(z_mean_acts, *z_mean);
+  El::Copy(z_log_sigma_acts, *z_log1);
+  AbsDistMat* z_log2 = z_log1->Copy();
+
+  //Compute entrywise add, square and exponent of the matrices
+  El::EntrywiseMap(*z_log1,El::MakeFunction(addone));
+  El::EntrywiseMap(*z_mean,El::MakeFunction(sq));
+  El::EntrywiseMap(*z_log2,El::MakeFunction(expn));
+
+  //Entrywise substraction of matrices
+  El::Axpy(DataType(-1),*z_mean,*z_log1);
+  El::Axpy(DataType(-1),*z_log2,*z_log1);
+  
+  //Entrywise mean of all the previous computation (stats)
+  entrywise_mean_and_stdev(*z_log1,mean_value,std_value);
+  return (0.5 * mean_value);
+}
+
+void kl_divergence::differentiate() {
+  const AbsDistMat& z_mean_acts = m_z_mean_layer->get_activations();
+  const AbsDistMat& z_log_sigma_acts = m_z_log_sigma_layer->get_activations();
+  auto height= z_mean_acts.Height();
+  
+  AbsDistMat* z_mean_gradient = z_mean_acts.Construct(z_mean_acts.Grid(),
+                                                      z_mean_acts.Root());
+  
+  AbsDistMat* z_log_sigma_gradient = z_log_sigma_acts.Construct(z_log_sigma_acts.Grid(),
+                                                     z_log_sigma_acts.Root());
+  El::Copy(z_mean_acts,*z_mean_gradient); 
+  El::Copy(z_log_sigma_acts,*z_log_sigma_gradient);
+
+  const DataType scale = DataType(1) / DataType(2)*height;
+  
+  //Compute z_log_sigma_gradient = -(1/2n)(1-exp(Z_logsigma))
+  auto expn  = [&](const DataType& x) {
+    return std::exp(x);
+  };
+  El::EntrywiseMap(*z_log_sigma_gradient,El::MakeFunction(expn));
+  El::Axpy(DataType(-1),*z_log_sigma_gradient, *z_log_sigma_gradient);
```

Comment:
This will result in a matrix full of zeros. You should change the `expn` lambda to something like `return 1 - std::exp(x)`.